sensor/common/registry/lazy_tls_image_registry.go

```diff
@@ -0,0 +1,176 @@
+	}
+
+	return &types.Config{
+		Username:         l.dockerConfig.GetUsername(),
+		Password:         l.dockerConfig.GetPassword(),
+		Insecure:         l.dockerConfig.GetInsecure(),
+		URL:              l.url,
+		RegistryHostname: l.registryHostname,
+	}
+}
+
+func (l *lazyTLSCheckRegistry) DataSource() *storage.DataSource {
+	return l.dataSource
+}
+
+func (l *lazyTLSCheckRegistry) HTTPClient() *http.Client {
+	utils.Should(errors.New("not implemented"))
+	return nil
+}
+
+func (l *lazyTLSCheckRegistry) Match(image *storage.ImageName) bool {
+	return urlfmt.TrimHTTPPrefixes(l.registryHostname) == image.GetRegistry()
+}
+
+func (l *lazyTLSCheckRegistry) Metadata(image *storage.Image) (*storage.ImageMetadata, error) {
+	// Attempt initialization since Metadata interacts with the registry.
+	l.lazyInit()
+
+	// initError and registry are modified while the
+	// write lock is held, to avoid a race grab the
+	// read lock.
+	l.initializedMutex.RLock()
+	defer l.initializedMutex.RUnlock()
+
+	if l.initError != nil {
+		return nil, l.initError
+	}
+
+	return l.registry.Metadata(image)
+}
+
+func (l *lazyTLSCheckRegistry) Name() string {
+	// source is modified while the write lock is held,
+	// to avoid a race grab the read lock.
+	l.initializedMutex.RLock()
+	defer l.initializedMutex.RUnlock()
+
+	return l.source.GetName()
+}
+
+func (l *lazyTLSCheckRegistry) Source() *storage.ImageIntegration {
+	// source is modified while the write lock is held,
+	// to avoid a race grab the read lock.
+	l.initializedMutex.RLock()
+	defer l.initializedMutex.RUnlock()
+
+	return l.source
+}
+
+func (l *lazyTLSCheckRegistry) Test() error {
+	utils.Should(errors.New("not implemented"))
+	return nil
+}
+
+// lazyInit attempts to lazily perform a TLS check and initialize the
+// underlying registry.  The concurrency mechanisms are based of
+// https://cs.opensource.google/go/go/+/refs/tags/go1.22.5:src/sync/once.go;l=48
+func (l *lazyTLSCheckRegistry) lazyInit() {
+	if l.isInitialized() {
+		return
+	}
+
+	l.initializedMutex.Lock()
+	defer l.initializedMutex.Unlock()
+
+	// Short-circuit if another goroutine completed initialization.
+	if l.isInitialized() {
+		return
+	}
+
+	// Do the TLS check.
+	secure, skip, err := l.tlsCheckCache.CheckTLS(context.Background(), l.dockerConfig.GetEndpoint())
+	if err != nil {
+		log.Warnf("Lazy TLS check failed for %q (%s): %v", l.source.GetName(), l.source.GetId(), err)
+		l.initError = err
+		return
+	}
+
+	if skip {
+		l.initError = errors.New("tls check skipped due to previous TLS check errors")
+		return
+	}
+
+	// If we got here assume that initialization has completed, any errors encountered
+	// are no longer temporal so do not try to repeat initialization.
+	defer l.initialized.Store(1)
+
+	// Clone the source to prevent a race.
+	newSrc := l.source.Clone()
+	dockerCfg, err := extractDockerConfig(newSrc)
+	if err != nil {
+		l.initError = err
+		return
+	}
+	dockerCfg.Insecure = !secure
+	l.source = newSrc
```

Comment:
does this not need to be protected?

Reply:
It is, that logic executed while `l.initializedMutex.Lock()` is held.

Reply:
in that case, why do we need to clone above? seems like writes are protected, right?

Reply:
It may not be needed given current usage throughout Sensor - however - without it a data race can occur when 'something' is reading the fields of the source/dockerConfig - the clone, IMHO, makes this more 'future proof' should usage change.

Without the clone running the unit tests will trigger a race occasionally:

```
=== RUN   TestAttemptToTriggerRace
==================
WARNING: DATA RACE
Write at 0x00c0019814d0 by goroutine 37:
  github.com/stackrox/rox/sensor/common/registry.(*lazyTLSCheckRegistry).lazyInit()
      /Users/dcaravel/dev/stackrox/stackrox-lazy-tls-check/sensor/common/registry/lazy_tls_image_registry.go:165 +0x22c
  github.com/stackrox/rox/sensor/common/registry.(*lazyTLSCheckRegistry).Metadata()
      /Users/dcaravel/dev/stackrox/stackrox-lazy-tls-check/sensor/common/registry/lazy_tls_image_registry.go:85 +0x38
  github.com/stackrox/rox/sensor/common/registry.TestAttemptToTriggerRace.func2()
      /Users/dcaravel/dev/stackrox/stackrox-lazy-tls-check/sensor/common/registry/lazy_tls_image_registry_test.go:224 +0x60
  github.com/stackrox/rox/sensor/common/registry.TestAttemptToTriggerRace.gowrap1()
      /Users/dcaravel/dev/stackrox/stackrox-lazy-tls-check/sensor/common/registry/lazy_tls_image_registry_test.go:227 +0x44

Previous read at 0x00c0019814d0 by goroutine 26:
  github.com/stackrox/rox/generated/storage.(*DockerConfig).GetInsecure()
      /Users/dcaravel/dev/stackrox/stackrox-lazy-tls-check/generated/storage/image_integration.pb.go:1018 +0xc8
  github.com/stackrox/rox/sensor/common/registry.TestAttemptToTriggerRace.func3()
      /Users/dcaravel/dev/stackrox/stackrox-lazy-tls-check/sensor/common/registry/lazy_tls_image_registry_test.go:235 +0xac
  github.com/stackrox/rox/sensor/common/registry.TestAttemptToTriggerRace.gowrap2()
      /Users/dcaravel/dev/stackrox/stackrox-lazy-tls-check/sensor/common/registry/lazy_tls_image_registry_test.go:237 +0x44

Goroutine 37 (running) created at:
  github.com/stackrox/rox/sensor/common/registry.TestAttemptToTriggerRace()
      /Users/dcaravel/dev/stackrox/stackrox-lazy-tls-check/sensor/common/registry/lazy_tls_image_registry_test.go:220 +0x710
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1689 +0x180
  testing.(*T).Run.gowrap1()
      /usr/local/go/src/testing/testing.go:1742 +0x40

Goroutine 26 (finished) created at:
  github.com/stackrox/rox/sensor/common/registry.TestAttemptToTriggerRace()
      /Users/dcaravel/dev/stackrox/stackrox-lazy-tls-check/sensor/common/registry/lazy_tls_image_registry_test.go:230 +0x86c
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1689 +0x180
  testing.(*T).Run.gowrap1()
      /usr/local/go/src/testing/testing.go:1742 +0x40
==================
common/registry: 2024/07/23 17:30:51.748782 /Users/dcaravel/dev/stackrox/stackrox-lazy-tls-check/sensor/common/registry/lazy_tls_image_registry.go:170: Warn: Lazy init failed for "fake-name" (fake-id), secure: true: too slow
    /Users/dcaravel/dev/stackrox/stackrox-lazy-tls-check/sensor/common/registry/testing.go:1398: race detected during execution of test
--- FAIL: TestAttemptToTriggerRace (2.71s)
FAIL
FAIL	github.com/stackrox/rox/sensor/common/registry	4.692s
```